# 4.8 Grouping Data & Aggregating Variables
## Group data in Python using groupby() function
## Use aggregation functions when deriving new columns
### -Find aggregated mean of 'order_number', grouped by 'department_id' for entire dataset
### -create 'loyalty_flag' using transform() and loc() functions
### -checking head of only selected columns
### -create 'spending_flag' based on average price across all their orders (low <10, high>=10)
### -create 'frequency_flag' that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column
### -turn outliers (all values over 100 dollars in the price column) into NaNs as this cleans data and we want it saved in this export (not just in the visualizations folder)
#### -export as ords_prods_grouped.pkl 

## Import Libraries and data

In [1]:
import pandas as pd
import numpy as np
import os

In [2]:
#import path
path = r'/Users/nancykray/Desktop/Instacart Basket Analysis'

In [3]:
path

'/Users/nancykray/Desktop/Instacart Basket Analysis'

In [4]:
#import dataframe 'ords_prods_merged_derived'
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merged_derived.pkl'))

## Shape & Head Check

In [5]:
#check the shape of dataframe
ords_prods_merge.shape

(32404859, 19)

In [6]:
#check the head (to look at columns & rows)
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_time  \
0   2539329        1             1                   2           8   
1   2398795        1             2                   3           7   
2    473747        1             3                   3          12   
3   2254736        1             4                   4           7   
4    431534        1             5                   4          15   

   days_since_prior_order      first_order  product_id  add_to_cart_order  \
0                     NaN      First Order         196                  1   
1                    15.0  Repeat Customer         196                  1   
2                    21.0  Repeat Customer         196                  1   
3                    29.0  Repeat Customer         196                  1   
4                    28.0  Repeat Customer         196                  1   

   reordered _merge product_name  aisle_id  department_id  prices  \
0          0   both         Soda        77              7     9.0   
1          1   both         Soda        77              7     9.0   
2          1   both         Soda        77              7     9.0   
3          1   both         Soda        77              7     9.0   
4          1   both         Soda        77              7     9.0   

     price_range_loc     busiest_day     busiest_days busiest_period_of_day  
0  Mid-range product  Regularly busy   Regularly busy        Average orders  
1  Mid-range product  Regularly busy  Least busy days        Average orders  
2  Mid-range product  Regularly busy  Least busy days           Most orders  
3  Mid-range product      Least busy  Least busy days        Average orders  
4  Mid-range product      Least busy  Least busy days           Most orders

## #2 Find the aggregated mean of 'order_number', grouped by 'department_id' for entire dataset

In [7]:
# Step 1 & 2 in the process: use groupby() and agg functions to calculate the mean of 'order_number'
ords_prods_merge.groupby('department_id').agg({'order_number':['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

## #3 Analyze the result.  How do these results differ for the entire dataframe compared to those of the subset?

##### ### The first noticable observation is that the 'department_id's are listed in order from 1-21, showing that it is complete for the entire dataset.  The subset only showed 8 rows/'department_id's (4, 7, 13, 14, 16, 17, 19, 20).
###### Another observation is that the subset only showed the mean values of lowest 11.2 to highest 19.46.  The entire dataframe has a max mean of 22 for 'department_id" 21.  Upon this observation, the lowest mean has changed, leading to the next point.
###### And finally, all of the means from the entire data set are lower in general than the subset, the the exception of only one 'department_id', that of 17.  This mean increased significantly from 11.29 (subset) to 15.69 (entire dataframe).  This is worth looking into as perhaps there are some outliers that affected the mean.

## #4 Follow the instructions in the Exercise for creating a loyalty flag for existing customers using the transform() and loc() functions.
### Part 1

In [8]:
# Find the loyal customers by using the transform() function to get new a new "max_order' column as follows:
ords_prods_merge['max_order']=ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

/var/folders/fb/2jrdrrb52mv7jbhxt0nbx_740000gn/T/ipykernel_15219/2234857839.py:2: FutureWarning: The provided callable <function max at 0x1100d9940> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  ords_prods_merge['max_order']=ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)


In [9]:
#check to see if the 'max_order' column is there
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_time  \
0   2539329        1             1                   2           8   
1   2398795        1             2                   3           7   
2    473747        1             3                   3          12   
3   2254736        1             4                   4           7   
4    431534        1             5                   4          15   

   days_since_prior_order      first_order  product_id  add_to_cart_order  \
0                     NaN      First Order         196                  1   
1                    15.0  Repeat Customer         196                  1   
2                    21.0  Repeat Customer         196                  1   
3                    29.0  Repeat Customer         196                  1   
4                    28.0  Repeat Customer         196                  1   

   reordered _merge product_name  aisle_id  department_id  prices  \
0          0   both         Soda        77              7     9.0   
1          1   both         Soda        77              7     9.0   
2          1   both         Soda        77              7     9.0   
3          1   both         Soda        77              7     9.0   
4          1   both         Soda        77              7     9.0   

     price_range_loc     busiest_day     busiest_days busiest_period_of_day  \
0  Mid-range product  Regularly busy   Regularly busy        Average orders   
1  Mid-range product  Regularly busy  Least busy days        Average orders   
2  Mid-range product  Regularly busy  Least busy days           Most orders   
3  Mid-range product      Least busy  Least busy days        Average orders   
4  Mid-range product      Least busy  Least busy days           Most orders   

   max_order  
0         10  
1         10  
2         10  
3         10  
4         10

In [10]:
#print a head of 100 rows to further check the data
ords_prods_merge.head(100)

order_id  user_id  order_number  orders_day_of_week  order_time  \
0    2539329        1             1                   2           8   
1    2398795        1             2                   3           7   
2     473747        1             3                   3          12   
3    2254736        1             4                   4           7   
4     431534        1             5                   4          15   
..       ...      ...           ...                 ...         ...   
95   3226575      360             1                   5          12   
96   1469869      377             3                   5          17   
97   1927023      387             2                   4          10   
98    858092      420             4                   1          19   
99    119314      420             5                   3          19   

    days_since_prior_order      first_order  product_id  add_to_cart_order  \
0                      NaN      First Order         196                  1   
1                     15.0  Repeat Customer         196                  1   
2                     21.0  Repeat Customer         196                  1   
3                     29.0  Repeat Customer         196                  1   
4                     28.0  Repeat Customer         196                  1   
..                     ...              ...         ...                ...   
95                     NaN      First Order         196                  1   
96                     3.0  Repeat Customer         196                  9   
97                    22.0  Repeat Customer         196                  3   
98                    30.0  Repeat Customer         196                  2   
99                    16.0  Repeat Customer         196                  2   

    reordered _merge product_name  aisle_id  department_id  prices  \
0           0   both         Soda        77              7     9.0   
1           1   both         Soda        77              7     9.0   
2           1   both         Soda        77              7     9.0   
3           1   both         Soda        77              7     9.0   
4           1   both         Soda        77              7     9.0   
..        ...    ...          ...       ...            ...     ...   
95          0   both         Soda        77              7     9.0   
96          0   both         Soda        77              7     9.0   
97          0   both         Soda        77              7     9.0   
98          0   both         Soda        77              7     9.0   
99          1   both         Soda        77              7     9.0   

      price_range_loc     busiest_day     busiest_days busiest_period_of_day  \
0   Mid-range product  Regularly busy   Regularly busy        Average orders   
1   Mid-range product  Regularly busy  Least busy days        Average orders   
2   Mid-range product  Regularly busy  Least busy days           Most orders   
3   Mid-range product      Least busy  Least busy days        Average orders   
4   Mid-range product      Least busy  Least busy days           Most orders   
..                ...             ...              ...                   ...   
95  Mid-range product  Regularly busy   Regularly busy           Most orders   
96  Mid-range product  Regularly busy   Regularly busy        Average orders   
97  Mid-range product      Least busy  Least busy days           Most orders   
98  Mid-range product  Regularly busy     Busiest days        Average orders   
99  Mid-range product  Regularly busy  Least busy days        Average orders   

    max_order  
0          10  
1          10  
2          10  
3          10  
4          10  
..        ...  
95          3  
96          3  
97          8  
98         22  
99         22  

[100 rows x 20 columns]

In [11]:
#change the limit so that we can see the 100 rows in the head function
pd.options.display.max_rows = None

In [12]:
ords_prods_merge.head(100)

order_id  user_id  order_number  orders_day_of_week  order_time  \
0    2539329        1             1                   2           8   
1    2398795        1             2                   3           7   
2     473747        1             3                   3          12   
3    2254736        1             4                   4           7   
4     431534        1             5                   4          15   
5    3367565        1             6                   2           7   
6     550135        1             7                   1           9   
7    3108588        1             8                   1          14   
8    2295261        1             9                   1          16   
9    2550362        1            10                   4           8   
10   2968173       15            15                   1           9   
11   1870022       15            17                   2          16   
12   1911383       15            18                   2          11   
13   2715276       15            21                   1           9   
14    487368       15            22                   1          10   
15   2293453       19             2                   5          14   
16   1973799       19             5                   6          12   
17    532817       19             7                   4          17   
18   1573906       21            10                   3          10   
19   1593000       31            10                   3           8   
20   2231262       31            17                   3          11   
21   2580647       43             6                   4          16   
22   2187180       43             9                   4          12   
23   2497897       52             1                   1           9   
24   1318871       52             2                   1          10   
25   1261384       52             3                   1          10   
26    580568       52             5                   2          10   
27   2428073       52             7                   2           7   
28    342306       52             9                   1          11   
29   1520353       52            10                   2          10   
30   2542086       52            11                   2           9   
31    944694       52            12                   1          16   
32   1268191       52            15                   2          11   
33   1122089       52            17                   2          10   
34   1498922       52            18                   2          10   
35    180919       52            19                   3           8   
36   2695875       52            21                   5          10   
37   3205960       67             1                   3          11   
38    970922       67             2                   1          15   
39    596585       67             3                   1          13   
40   2037087       67             4                   2           9   
41   2474111       67             5                   3          14   
42   1568481       67             6                   3          12   
43   2048527       67             8                   2          11   
44   2389031       67             9                   2          14   
45   2378562       67            10                   1          10   
46   1515698       67            11                   1          11   
47   1928050       67            12                   2          12   
48   1949274       67            13                   1          10   
49    551869       67            14                   1           9   
50    768572       67            16                   4          10   
51    150528       67            18                   5          16   
52     29758       67            19                   0          14   
53   1558976       67            21                   1           9   
54    863974       67            22                   1          11   
55   2308399       67       

In [13]:
# cleared output after check to save space

### Part 2: use loc( ) function to create loyalty label to user ID based on their 'max_order' value based on criteria:
#### LOYAL CUSTOMER -> MAX ORDER >40
#### REGULAR CUSTOMER -> MAX ORDER > 10, BUT <= 40
#### NEW CUSTOMER -> MAX ORDER <= 10


In [14]:
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal Customer'

In [15]:
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order']>10),'loyalty_flag'] = 'Regular Customer'

In [16]:
ords_prods_merge.loc[ords_prods_merge['max_order'] <=10, 'loyalty_flag'] = 'New Customer'

In [17]:
#check frequency of values in 'loyalty_flag' column
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular Customer    15876776
Loyal Customer      10284093
New Customer         6243990
Name: count, dtype: int64

In [18]:
#check dataframe again with new column added, 'loyalty_flag'
ords_prods_merge.head(50)

order_id  user_id  order_number  orders_day_of_week  order_time  \
0    2539329        1             1                   2           8   
1    2398795        1             2                   3           7   
2     473747        1             3                   3          12   
3    2254736        1             4                   4           7   
4     431534        1             5                   4          15   
5    3367565        1             6                   2           7   
6     550135        1             7                   1           9   
7    3108588        1             8                   1          14   
8    2295261        1             9                   1          16   
9    2550362        1            10                   4           8   
10   2968173       15            15                   1           9   
11   1870022       15            17                   2          16   
12   1911383       15            18                   2          11   
13   2715276       15            21                   1           9   
14    487368       15            22                   1          10   
15   2293453       19             2                   5          14   
16   1973799       19             5                   6          12   
17    532817       19             7                   4          17   
18   1573906       21            10                   3          10   
19   1593000       31            10                   3           8   
20   2231262       31            17                   3          11   
21   2580647       43             6                   4          16   
22   2187180       43             9                   4          12   
23   2497897       52             1                   1           9   
24   1318871       52             2                   1          10   
25   1261384       52             3                   1          10   
26    580568       52             5                   2          10   
27   2428073       52             7                   2           7   
28    342306       52             9                   1          11   
29   1520353       52            10                   2          10   
30   2542086       52            11                   2           9   
31    944694       52            12                   1          16   
32   1268191       52            15                   2          11   
33   1122089       52            17                   2          10   
34   1498922       52            18                   2          10   
35    180919       52            19                   3           8   
36   2695875       52            21                   5          10   
37   3205960       67             1                   3          11   
38    970922       67             2                   1          15   
39    596585       67             3                   1          13   
40   2037087       67             4                   2           9   
41   2474111       67             5                   3          14   
42   1568481       67             6                   3          12   
43   2048527       67             8                   2          11   
44   2389031       67             9                   2          14   
45   2378562       67            10                   1          10   
46   1515698       67            11                   1          11   
47   1928050       67            12                   2          12   
48   1949274       67            13                   1          10   
49    551869       67            14                   1           9   

    days_since_prior_order      first_order  product_id  add_to_cart_order  \
0                      NaN      First Order         196                  1   
1                     15.0  Repeat Customer         196                  1   
2                     21.0  Repeat Customer         196                  1   
3                     29.0  Repeat Customer         196               

In [19]:
#check the head function for only select columns
ords_prods_merge[['user_id', 'loyalty_flag', 'order_number']].head(60)

user_id      loyalty_flag  order_number
0         1      New Customer             1
1         1      New Customer             2
2         1      New Customer             3
3         1      New Customer             4
4         1      New Customer             5
5         1      New Customer             6
6         1      New Customer             7
7         1      New Customer             8
8         1      New Customer             9
9         1      New Customer            10
10       15  Regular Customer            15
11       15  Regular Customer            17
12       15  Regular Customer            18
13       15  Regular Customer            21
14       15  Regular Customer            22
15       19      New Customer             2
16       19      New Customer             5
17       19      New Customer             7
18       21  Regular Customer            10
19       31  Regular Customer            10
20       31  Regular Customer            17
21       43  Regular Customer             6
22       43  Regular Customer             9
23       52  Regular Customer             1
24       52  Regular Customer             2
25       52  Regular Customer             3
26       52  Regular Customer             5
27       52  Regular Customer             7
28       52  Regular Customer             9
29       52  Regular Customer            10
30       52  Regular Customer            11
31       52  Regular Customer            12
32       52  Regular Customer            15
33       52  Regular Customer            17
34       52  Regular Customer            18
35       52  Regular Customer            19
36       52  Regular Customer            21
37       67  Regular Customer             1
38       67  Regular Customer             2
39       67  Regular Customer             3
40       67  Regular Customer             4
41       67  Regular Customer             5
42       67  Regular Customer             6
43       67  Regular Customer             8
44       67  Regular Customer             9
45       67  Regular Customer            10
46       67  Regular Customer            11
47       67  Regular Customer            12
48       67  Regular Customer            13
49       67  Regular Customer            14
50       67  Regular Customer            16
51       67  Regular Customer            18
52       67  Regular Customer            19
53       67  Regular Customer            21
54       67  Regular Customer            22
55       67  Regular Customer            23
56       81      New Customer             3
57       81      New Customer             5
58       82  Regular Customer             1
59       82  Regular Customer             3

## #5 The marketing team at Instacart wants to know whether there’s a difference between the spending habits of the three types of customers you identified. Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category (Loyal Customer, Regular Customer, and New Customer). What you’re trying to determine is whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.

In [20]:
# use groupby function to split into groups based on spending habits of each of the three loyalty groups
ords_prods_merge.groupby('loyalty_flag').agg({'prices' :['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal Customer    10.386336  1.0  99999.0
New Customer      13.294670  1.0  99999.0
Regular Customer  12.495717  1.0  99999.0

#### the loyal customers tend to spend slightly less per product than regular and new customers by 2-3 dollars

## #6 The team now wants to target different types of spenders in their marketing campaigns. This can be achieved by looking at the prices of the items people are buying. Create a spending flag for each user based on the average price across all their orders using the following criteria:
If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender.”

If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender.”


### Part 1

In [21]:
#use the transform() function for creating a 'mean_product_price' column for each user id
ords_prods_merge['mean_product_price']=ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

/var/folders/fb/2jrdrrb52mv7jbhxt0nbx_740000gn/T/ipykernel_15219/76109492.py:2: FutureWarning: The provided callable <function mean at 0x1100da340> is currently using SeriesGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "mean" instead.
  ords_prods_merge['mean_product_price']=ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)


In [22]:
#look at the columns of interest only using head
ords_prods_merge[['order_id', 'mean_product_price', 'prices']].head(10)

order_id  mean_product_price  prices
0   2539329            6.367797     9.0
1   2398795            6.367797     9.0
2    473747            6.367797     9.0
3   2254736            6.367797     9.0
4    431534            6.367797     9.0
5   3367565            6.367797     9.0
6    550135            6.367797     9.0
7   3108588            6.367797     9.0
8   2295261            6.367797     9.0
9   2550362            6.367797     9.0

### Part 2

In [23]:
#If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender.”
ords_prods_merge.loc[ords_prods_merge['mean_product_price'] < 10, 'spending_flag'] = 'Low Spender'

In [24]:
#If the mean of the prices of products purchased by a user is higher than or equal to 10, flag as “High spender.”
ords_prods_merge.loc[ords_prods_merge['mean_product_price'] >=10, 'spending_flag'] = 'High Spender'

In [25]:
#check the frequency of values in the new 'spending_flag' column using value_counts function
ords_prods_merge['spending_flag'].value_counts(dropna=False)

spending_flag
Low Spender     31770614
High Spender      634245
Name: count, dtype: int64

## #7 In order to send relevant notifications to users within the app (for instance, asking users if they want to buy the same item again), the Instacart team wants you to determine frequent versus non-frequent customers. Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column. The criteria for the flag should be as follows:
##### If the median of “days_since_prior_order” is higher than 20, then the customer should be labeled a “Non-frequent customer.”
##### If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a “Regular customer.”
##### If the median is lower than or equal to 10, then the customer should be labeled a “Frequent customer.”


### Part 1

In [26]:
#create a column using .loc() function
ords_prods_merge['customer_frequency']=ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

/var/folders/fb/2jrdrrb52mv7jbhxt0nbx_740000gn/T/ipykernel_15219/1549844515.py:2: FutureWarning: The provided callable <function median at 0x110226de0> is currently using SeriesGroupBy.median. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "median" instead.
  ords_prods_merge['customer_frequency']=ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)


In [27]:
#check to see the column in the dataset
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_time  \
0   2539329        1             1                   2           8   
1   2398795        1             2                   3           7   
2    473747        1             3                   3          12   
3   2254736        1             4                   4           7   
4    431534        1             5                   4          15   

   days_since_prior_order      first_order  product_id  add_to_cart_order  \
0                     NaN      First Order         196                  1   
1                    15.0  Repeat Customer         196                  1   
2                    21.0  Repeat Customer         196                  1   
3                    29.0  Repeat Customer         196                  1   
4                    28.0  Repeat Customer         196                  1   

   reordered  ... prices    price_range_loc     busiest_day     busiest_days  \
0          0  ...    9.0  Mid-range product  Regularly busy   Regularly busy   
1          1  ...    9.0  Mid-range product  Regularly busy  Least busy days   
2          1  ...    9.0  Mid-range product  Regularly busy  Least busy days   
3          1  ...    9.0  Mid-range product      Least busy  Least busy days   
4          1  ...    9.0  Mid-range product      Least busy  Least busy days   

   busiest_period_of_day max_order  loyalty_flag mean_product_price  \
0         Average orders        10  New Customer           6.367797   
1         Average orders        10  New Customer           6.367797   
2            Most orders        10  New Customer           6.367797   
3         Average orders        10  New Customer           6.367797   
4            Most orders        10  New Customer           6.367797   

  spending_flag  customer_frequency  
0   Low Spender                20.5  
1   Low Spender                20.5  
2   Low Spender                20.5  
3   Low Spender                20.5  
4   Low Spender                20.5  

[5 rows x 24 columns]

### Part 2

In [28]:
#If median of “days_since_prior_order” is higher than 20, then customer should be labeled “Non-frequent customer.”
ords_prods_merge.loc[ords_prods_merge['customer_frequency']  > 20, 'frequency_flag'] = 'Non-frequent Customer'

In [29]:
#If the median >10 and < =20, then customer should be labeled “Regular customer#
ords_prods_merge.loc[(ords_prods_merge['customer_frequency'] <= 20) & (ords_prods_merge ['customer_frequency']>10),'frequency_flag'] = 'Regular Customer'

In [30]:
#If the median is <=10, then the customer should be labeled a “Frequent customer.”
ords_prods_merge.loc[ords_prods_merge['customer_frequency'] <=10, 'frequency_flag'] = 'Frequent Customer'

In [31]:
#check the frequency flag column in value_counts
ords_prods_merge['frequency_flag'].value_counts(dropna=False)

frequency_flag
Frequent Customer        21559853
Regular Customer          7208564
Non-frequent Customer     3636437
nan                             5
Name: count, dtype: int64

In [32]:
#check the relevant columns using head function
ords_prods_merge[['user_id', 'customer_frequency', 'frequency_flag']].head(60)

user_id  customer_frequency         frequency_flag
0         1                20.5  Non-frequent Customer
1         1                20.5  Non-frequent Customer
2         1                20.5  Non-frequent Customer
3         1                20.5  Non-frequent Customer
4         1                20.5  Non-frequent Customer
5         1                20.5  Non-frequent Customer
6         1                20.5  Non-frequent Customer
7         1                20.5  Non-frequent Customer
8         1                20.5  Non-frequent Customer
9         1                20.5  Non-frequent Customer
10       15                10.0      Frequent Customer
11       15                10.0      Frequent Customer
12       15                10.0      Frequent Customer
13       15                10.0      Frequent Customer
14       15                10.0      Frequent Customer
15       19                 6.0      Frequent Customer
16       19                 6.0      Frequent Customer
17       19                 6.0      Frequent Customer
18       21                 7.5      Frequent Customer
19       31                 6.0      Frequent Customer
20       31                 6.0      Frequent Customer
21       43                10.0      Frequent Customer
22       43                10.0      Frequent Customer
23       52                 8.0      Frequent Customer
24       52                 8.0      Frequent Customer
25       52                 8.0      Frequent Customer
26       52                 8.0      Frequent Customer
27       52                 8.0      Frequent Customer
28       52                 8.0      Frequent Customer
29       52                 8.0      Frequent Customer
30       52                 8.0      Frequent Customer
31       52                 8.0      Frequent Customer
32       52                 8.0      Frequent Customer
33       52                 8.0      Frequent Customer
34       52                 8.0      Frequent Customer
35       52                 8.0      Frequent Customer
36       52                 8.0      Frequent Customer
37       67                 7.0      Frequent Customer
38       67                 7.0      Frequent Customer
39       67                 7.0      Frequent Customer
40       67                 7.0      Frequent Customer
41       67                 7.0      Frequent Customer
42       67                 7.0      Frequent Customer
43       67                 7.0      Frequent Customer
44       67                 7.0      Frequent Customer
45       67                 7.0      Frequent Customer
46       67                 7.0      Frequent Customer
47       67                 7.0      Frequent Customer
48       67                 7.0      Frequent Customer
49       67                 7.0      Frequent Customer
50       67                 7.0      Frequent Customer
51       67                 7.0      Frequent Customer
52       67                 7.0      Frequent Customer
53       67                 7.0      Frequent Customer
54       67                 7.0      Frequent Customer
55       67                 7.0      Frequent Customer
56       81                30.0  Non-frequent Customer
57       81                30.0  Non-frequent Customer
58       82                11.5       Regular Customer
59       82                11.5       Regular Customer

In [1]:
#check shape again before exporting
ords_prods_merge.shape

NameError: name 'ords_prods_merge' is not defined

In [35]:
# As prompted from lesson 4.9 visualizations: 
# turn outliers (all values over 100 dollars in the price column) into NaNs as this cleans data 
# and we want it saved in this export (not just in the visualizations folder)
ords_prods_merge.loc[ords_prods_merge['prices'] >100, 'prices'] = np.nan

In [37]:
#just checking I get the same outcome I got in 4.9 
ords_prods_merge['prices'].max()

25.0

## Export data frame, 'ords_prods_grouped' as a pickle into 'Prepared Folder'

In [38]:
ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_grouped.pkl'))